include/deal.II/rol/vector_adaptor.h

```diff
@@ -0,0 +1,482 @@
+      Teuchos::dyn_cast<const VectorAdaptor>(rol_vector);
+
+    *vector_ptr += *(vector_adaptor.getVector());
+  }
+
+
+
+  template<typename VectorType>
+  void
+  VectorAdaptor<VectorType>::axpy (const value_type               alpha,
+                                   const ROL::Vector<value_type> &rol_vector)
+  {
+    Assert (this->dimension() == rol_vector.dimension(),
+            ExcDimensionMismatch(this->dimension(), rol_vector.dimension()));
+
+    const VectorAdaptor &vector_adaptor =
+      Teuchos::dyn_cast<const VectorAdaptor>(rol_vector);
+
+    vector_ptr->add (alpha, *(vector_adaptor.getVector()));
+  }
+
+
+
+  template<typename VectorType>
+  int
+  VectorAdaptor<VectorType>::dimension () const
+  {
+    Assert (vector_ptr->size() < std::numeric_limits<int>::max(),
+            ExcMessage("The size of the vector being used is greater than "
+                       "largest value of type int."));
+    return static_cast<int>(vector_ptr->size());
+  }
+
+
+
+  template<typename VectorType>
+  void
+  VectorAdaptor<VectorType>::scale (const value_type alpha)
+  {
+    (*vector_ptr) *= alpha;
+  }
+
+
+
+  template<typename VectorType>
+  typename VectorType::value_type
+  VectorAdaptor<VectorType>::
+  dot (const ROL::Vector<value_type> &rol_vector) const
+  {
+    Assert (this->dimension() == rol_vector.dimension(),
+            ExcDimensionMismatch(this->dimension(), rol_vector.dimension()));
+
+    const VectorAdaptor &vector_adaptor =
+      Teuchos::dyn_cast< const VectorAdaptor>(rol_vector);
+
+    return (*vector_ptr) * (*vector_adaptor.getVector());
+  }
+
+
+
+  template<typename VectorType>
+  typename VectorType::value_type
+  VectorAdaptor<VectorType>::norm() const
+  {
+    return internal::NumberType<value_type>::value(vector_ptr->l2_norm());
+  }
+
+
+
+  template<typename VectorType>
+  Teuchos::RCP<ROL::Vector<typename VectorType::value_type> >
+  VectorAdaptor<VectorType>::clone() const
+  {
+    Teuchos::RCP< VectorType> vec_ptr = Teuchos::rcp (new VectorType);
+    (*vec_ptr) = (*vector_ptr);
+
+    return Teuchos::rcp (new VectorAdaptor (vec_ptr));
+  }
+
+
+
+  template<typename VectorType>
+  Teuchos::RCP<ROL::Vector<typename VectorType::value_type> >
+  VectorAdaptor<VectorType>::basis (const int i) const
+  {
+    Teuchos::RCP<VectorType> vec_ptr = Teuchos::rcp (new VectorType);
+
+    // Zero all the entries in dealii vector.
+    vec_ptr->reinit(*vector_ptr, false);
+
+    if (vector_ptr->locally_owned_elements().is_element(i))
+      vec_ptr->operator[](i) = 1.;
+
+    vec_ptr->compress(VectorOperation::insert);
+
+    Teuchos::RCP<VectorAdaptor> e = Teuchos::rcp (new VectorAdaptor(vec_ptr));
+
+    return e;
+  }
+
+
+
+  template<typename VectorType>
+  void
+  VectorAdaptor<VectorType>::
+  applyUnary (const ROL::Elementwise::UnaryFunction<value_type> &f)
+  {
+    for (typename VectorType::iterator
+         iterator  = vector_ptr->begin();
+         iterator != vector_ptr->end();
```

Comment:
This is slightly inefficient because you're going to call `end()` once for each element. It may be more efficient to store the end iterator before the loop starts.

But more importantly, for parallel vectors, this isn't going to work. You'll have to restrict the call to those elements that are locally owned. (If that's too cumbersome right now, then just put an `Assert(..., ExcNotImplemented())` into the function to make sure we get a decent error message if anyone ever encounters this.)

...hm, upon reading up on this, `begin/end` actually only iterate over the locally owned elements. Maybe prefix the loop by a comment that says "Iterate over the locally owned elements" to remind readers such as me that that's exactly what's happening here.

Reply:
Wouldn't most compilers figure that it should store the end iterator?

I will add the comment that we are iterating over locally owned elements.

Reply:
They can't because the compiler can, in general, not figure out that you're not changing `*vector_ptr` in the body of the loop. And if you change the vector pointed to in the body of the loop, `end()` may return a different value next time around.

(There are some cases where the compiler can convince itself that that's not the case, specifically if `end()` and all functions you call in the body are (i) inlined and (ii) simple enough for a compiler to analyze. But here, you're potentially iterating over index sets of locally owned elements and do something on a functor whose action is not known to the compiler, so I think you'll be out of luck.)

Reply:
Thank you for clarifying!